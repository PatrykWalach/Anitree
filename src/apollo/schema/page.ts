import { Media, MediaType } from './media'
import { ScoreFormat } from './viewer'

export interface Variables {
  idIn?: number[] | number
  page?: number
  format?: ScoreFormat
  isAdult?: boolean
  onList?: boolean
  search?: string
  includedTags?: string[]
  year?: number
  type?: MediaType
  season?: MediaSeason
  sort?: MediaSort[]
}

export type MediaSeason = 'WINTER'
export type MediaSort = 'POPULARITY_DESC'

export interface PageInfo {
  total: number
  hasNextPage: boolean
  lastPage: number
  currentPage: number
}

export interface Page {
  pageInfo: PageInfo
  media: Media[]
}
