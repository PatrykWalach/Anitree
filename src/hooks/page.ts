import { useQuery, useResult, UseQueryOptions } from '@vue/apollo-composable'
import { useComputedOrCallback } from './reducer'
import { Ref, ref } from '@vue/composition-api'
import { DocumentNode } from 'graphql'
import { Page_page } from './__generated__/Page_page'
import { Page_media } from './__generated__/Page_media'

interface Page extends Page_page {
  media: readonly (Page_media | null)[] | null
}

export const usePage = <R extends { Page: Page | null }, V>(
  queryDocument: DocumentNode,
  value: Ref<V> | (() => V),
  options:
    | UseQueryOptions<R, V>
    | Ref<UseQueryOptions<R, V>>
    | (() => UseQueryOptions<R, V>),
) => {
  const variables = useComputedOrCallback(value)
  const query = useQuery<R, V>(queryDocument, variables, options)

  const currentPage = useResult(
    query.result,
    0,
    data => data?.Page?.pageInfo?.currentPage || 0,
  )

  const hasNextPage = useResult(
    query.result,
    false,
    data => data?.Page?.pageInfo?.hasNextPage || false,
  )
  const loadingMore = ref(false)

  const loadMore = async () => {
    if (!query.loading.value && !loadingMore.value && hasNextPage.value) {
      loadingMore.value = true
      await query.fetchMore({
        updateQuery: updatePageQuery,
        variables: {
          ...variables.value,
          page: 1 + currentPage.value,
        },
      })
      loadingMore.value = false
    }
  }

  return {
    ...query,
    loadMore,
    loadingMore,
    currentPage,
    hasNextPage,
  }
}

export const updatePageQuery = <P extends { Page: Page | null }>(
  previousResult: P,
  {
    fetchMoreResult,
  }: {
    fetchMoreResult?: P
  },
): P => {
  if (!fetchMoreResult) {
    return previousResult
  }

  const getPageAndMedia = (result: P) => {
    const resultPage = result.Page
    const resultMedia = resultPage?.media || []
    return [resultPage, resultMedia] as const
  }

  const [previousPage, previousMedia] = getPageAndMedia(previousResult)
  const [morePage, moreMedia] = getPageAndMedia(fetchMoreResult)

  return {
    ...previousResult,
    ...fetchMoreResult,
    Page: {
      __typename: 'Page',
      ...previousPage,
      ...morePage,
      media: previousMedia.concat(moreMedia),
    },
  }
}
