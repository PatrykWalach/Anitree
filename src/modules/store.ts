import * as modules from '.'
import { Store, plugin } from 'vuex-composition-api'

export const store =
  process.env.NODE_ENV === 'production'
    ? undefined
    : new Store({
        plugins: [plugin(Object.values(modules))],
      })
