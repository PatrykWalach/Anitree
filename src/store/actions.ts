export { titleActions as title } from './modules/title'

export { editActions as edit } from './modules/edit'
export { filterActions as filter } from './modules/filter'
export { shareActions as share } from './modules/share'
export { settingsActions as settings } from './modules/settings'
export { timelineActions as timeline } from './modules/timeline'
export { changesActions as changes } from './modules/changes'
