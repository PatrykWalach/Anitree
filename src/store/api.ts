import gql from 'graphql-tag'
import { FetchVariables } from './models'

import apollo from '../apollo'
import { AMedia } from '@/types'

export const query = gql`
  query($search: String, $idIn: [Int], $page: Int = 0) {
    Page(page: $page, perPage: 50) {
      media(search: $search, id_in: $idIn) {
        id
        bannerImage
        coverImage {
          extraLarge
          large
          medium
          color
        }
        format
        status
        tags {
          name
          id
          rank
        }
        siteUrl
        type
        studios(isMain: true) {
          nodes {
            name
            id
          }
        }
        startDate {
          year
          month
          day
        }
        endDate {
          year
          month
          day
        }
        title {
          romaji
          english
          native
        }
        description
        mediaListEntry {
          status
        }
        relations {
          edges {
            node {
              id
              type
            }
            id
            relationType
          }
        }
      }
    }
  }
`

export function fetchMediaApollo(variables: FetchVariables): Promise<AMedia[]> {
  return apollo
    .query<{ Page: { media: AMedia[] } }>({
      query,
      variables
    })
    .then(({ data }) => data.Page.media)
}
