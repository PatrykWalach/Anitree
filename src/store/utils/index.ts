export * from './useCommands'
export * from './useTitle'
export * from './useEdit'
