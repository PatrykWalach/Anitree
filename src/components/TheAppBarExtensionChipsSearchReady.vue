<template>
  <v-chip-group :style="{ 'max-width': '100%' }" :active-class="activeClass">
    <v-chip
      v-for="{ title, id } in media"
      :key="id"
      exact
      :to="{
        name: 'search',
        query: {
          search: title,
        },
      }"
    >
      {{ title }}
    </v-chip>
  </v-chip-group>
</template>
<script lang="ts">
import { computed, createComponent } from '@vue/composition-api'
import { Page } from '../graphql/schema/page'
import { useInjectedTheme } from './TheMediaAboutStats.vue'
import { useTitle } from '../store'

export interface Props {
  page: Page
}

export const useExtensionChip = () => {
  const { theme } = useInjectedTheme()

  const activeClass = computed(() =>
    theme.isDark ? 'white black--text ' : 'grey darken-2 white--text',
  )
  return {
    activeClass,
    theme,
  }
}

export default createComponent<Readonly<Props>>({
  props: {
    page: { default: null, required: true, type: Object },
  },
  setup(props) {
    const { getTitle } = useTitle()

    const media = computed(() =>
      props.page.media
        .map(({ id, title }) => ({ id, title: getTitle.value(title) }))
        .filter(
          ({ title }, i, arr) =>
            arr.findIndex(media => media.title === title) === i,
        ),
    )

    return {
      ...useExtensionChip(),
      media,
    }
  },
})
</script>
