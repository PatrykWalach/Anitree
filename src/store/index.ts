import Vue from 'vue'
import Vuex from 'vuex'

// import { getModule } from 'vuex-module-decorators'
Vue.use(Vuex)

import { ModuleMedia } from './modules/media'
import { ModuleTheme } from './modules/theme'
import { ModuleTitle } from './modules/title'
import { ModuleAuth } from './modules/auth'

export interface RootState {
  media: ModuleMedia
  theme: ModuleTheme
  title: ModuleTitle
  auth: ModuleAuth
}

export default new Vuex.Store<RootState>({})
