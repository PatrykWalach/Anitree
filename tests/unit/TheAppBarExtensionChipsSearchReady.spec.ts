import TheAppBarExtensionChipsSearchReady from '@/components/TheAppBarExtensionChipsSearchReady.vue'

import { matchSnapshot } from './utils'
import { mockedPageAndMedia } from './mocks/page'
import { useMockedStore } from './mocks/store'

describe('TheAppBarExtensionChipsSearchReady', () => {
  matchSnapshot(TheAppBarExtensionChipsSearchReady, {
    propsData: {
      page: mockedPageAndMedia,
    },
    provide: { ...useMockedStore() },
  })
})
