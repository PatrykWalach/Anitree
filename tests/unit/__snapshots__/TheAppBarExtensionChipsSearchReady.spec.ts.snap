// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TheAppBarExtensionChipsSearchReady matches snapshot 1`] = `
<v-chip-group-stub
  activeclass="grey darken-2 white--text"
  mobilebreakpoint="1264"
  nexticon="$next"
  previcon="$prev"
  style="max-width: 100%;"
>
  <v-chip-stub
    active="true"
    activeclass=""
    closeicon="$delete"
    exact="true"
    filtericon="$complete"
    tag="span"
    to="[object Object]"
  >
    
    Cowboy Bebop
  
  </v-chip-stub>
</v-chip-group-stub>
`;
